olclient/entity_helpers/work.py

```diff
@@ -0,0 +1,172 @@
+            }
+            # TODO: where is this statement appending to?
+            self.authors.append(author_role)
+            return author_role
+
+        def add_bookcover(self, url):
+            return self.OL.session.post(
+                f'{self.OL.base_url}/works/{self.olid}/-/add-cover',
+                files={'file': '', 'url': url, 'upload': 'submit'},
+            )
+
+        def add_subject(self, subject, comment=''):
+            return self.add_subjects([subject], comment)
+
+        def add_subjects(self, subjects, comment=''):
+            url = self.OL.base_url + "/works/" + self.olid + ".json"
+            data = self.OL.session.get(url).json()
+            original_subjects = data.get('subjects', [])
+            changed_subjects = merge_unique_lists([original_subjects, subjects])
+            data['_comment'] = comment or (
+                'adding %s to subjects' % ', '.join(subjects)
+            )
+            data['subjects'] = changed_subjects
+            return self.OL.session.put(url, json.dumps(data))
+
+        def rm_subjects(self, subjects, comment=''):
+            url = self.OL.base_url + "/works/" + self.olid + ".json"
+            r = self.OL.session.get(url)
+            data = r.json()
+            data['_comment'] = comment or ('rm subjects: %s' % ', '.join(subjects))
+            data['subjects'] = list(set(data['subjects']) - set(subjects))
+            return self.OL.session.put(url, json.dumps(data))
+
+        def save(self, comment):
+            body = self.json()
+            body['_comment'] = comment
+            url = self.OL.base_url + '/works/%s.json' % self.olid
+            return self.OL.session.put(url, json.dumps(body))
+
+        @classmethod
+        def get(cls, olid: str) -> Work:
+            path = '/works/%s.json' % olid
+            r = cls.OL.get_ol_response(path)
+            return cls(olid, **r.json())
+
+        @classmethod
+        def search(cls, title=None, author=None) -> Optional[Book]:
+            """Get the *closest* matching result in OpenLibrary based on a title
+            and author.
+
+            FIXME: This is essentially a Work and should be moved there
+            """
+            if not (title or author):
+                raise ValueError("Author or title required for metadata search")
+
+            url = f'{cls.OL.base_url}/search.json?title={title}'
+            if author:
+                url += '&author=%s' % author
+
+            @backoff.on_exception(
+                on_giveup=lambda error: logger.exception(
+                    "Error retrieving metadata for book: %s", error
+                ),
+                **cls.OL.BACKOFF_KWARGS,
+            )
+            def _get_book_by_metadata(ol_url):
+                return cls.OL.session.get(ol_url)
+
+            response = _get_book_by_metadata(url)
+
+            try:
+                results = Results(**response.json())
+            except Exception as e:
```

Comment:
https://realpython.com/the-most-diabolical-python-antipattern/

Reply:
commented explanation above, added the refactor to the same PR too. 👍🏽